Analysis of Ebay used cars listing

In this guided project, we'll work with a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website.
Out of the whole data, I have sampled 50,000 data points from the full dataset.
The main focus of this mini-project is to clean the data and analyze the included used car listings.


In [ ]:
import pandas as pd
import numpy as np

In [7]:
autos=pd.read_csv('autos.csv')
autos.head(10)

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_T�V_neu_ist_sehr_gepfleg...   
5  2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6  2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7  2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8  2016-03-22 16:51:34                                         Seat_Arosa   
9  2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   
5  privat   Angebot  $7,900     test         bus                2006   
6  privat   Angebot    $300     test   limousine                1995   
7  privat   Angebot  $1,990  control   limousine                1998   
8  privat   Angebot    $250     test         NaN                2000   
9  privat   Angebot    $590  control         bus                1997   

     gearbox  powerPS    model   odometer  monthOfRegistration fuelType  \
0    manuell      158   andere  150,000km                    3      lpg   
1  automatik      286      7er  150,000km                    6   benzin   
2    manuell      102     golf   70,000km                    7   benzin   
3  automatik       71   fortwo   70,000km                    6   benzin   
4    manuell        0    focus  150,000km                    7   benzin   
5  automatik      150  voyager  150,000km                    4   diesel   
6    manuell       90     golf  150,000km                    8   benzin   
7    manuell       90     golf  150,000km                   12   diesel   
8    manuell        0    arosa  150,000km                   10      NaN   
9    manuell       90   megane  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   
5    chrysler               NaN  2016-03-21 00:00:00             0   
6  volkswagen               NaN  2016-03-20 00:00:00             0   
7  volkswagen              nein  2016-03-16 00:00:00             0   
8        seat              nein  2016-03-22 00:00:00             0   
9     renault              nein  2016-03-16 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50  
5       22962  2016-04-06 09:45:21  
6       31535  2016-03-23 02:48:59  
7       53474  2016-04-07 03:17:32  
8        7426  2016-03-26 18:18:10  
9       15749  2016-04-06 10:46:35

In [6]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

The dataset contains 20 columns, most of which are strings.
The column names use camelcase instead of Python's preferred snakecase, which means we can't just replace spaces with underscores. We will change camelcase to snakecase. I am going to clean the data.
 
 1, Change column names to snakecase and rename some columns
 2, Remove some columns which have no values
 3, Change datatype to int for some colunmns so that I can do analysis later;Remove outliers
 4, Format date columns
 

In [8]:
#Change column names to snakecase and rename some columns
autos.columns

Index([u'dateCrawled', u'name', u'seller', u'offerType', u'price', u'abtest',
       u'vehicleType', u'yearOfRegistration', u'gearbox', u'powerPS', u'model',
       u'odometer', u'monthOfRegistration', u'fuelType', u'brand',
       u'notRepairedDamage', u'dateCreated', u'nrOfPictures', u'postalCode',
       u'lastSeen'],
      dtype='object')

In [10]:
autos.rename(columns={'yearOfRegistration' : 'registration_year',
                        'monthOfRegistration' : 'registration_month',
                        'notRepairedDamage' : 'unrepaired_damage',
                        'dateCreated' : 'ad_created',
                        'dateCrawled' : 'date_crawled', 
                        'offerType' : 'offer_type',
                        'abtest' : 'ab_test',
                        'vehicleType' : 'vehicle_type',
                        'powerPS' : 'power_ps',
                        'fuelType' : 'fuel_type',
                        'nrOfPictures' : 'num_photos',
                        'postalCode' : 'postal_code',
                        'lastSeen' : 'last_seen',
                     'odometer':'odometer_km'},inplace=True)
autos.columns

Index([u'date_crawled', u'name', u'seller', u'offer_type', u'price',
       u'ab_test', u'vehicle_type', u'registration_year', u'gearbox',
       u'power_ps', u'model', u'odometer_km', u'registration_month',
       u'fuel_type', u'brand', u'unrepaired_damage', u'ad_created',
       u'num_photos', u'postal_code', u'last_seen'],
      dtype='object')

In [11]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-30 17:37:35  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

       odometer_km  registration_month fuel_type       brand  \
count        50000        50000.000000     45518       50000   
unique          13                 NaN         7          40   
top      150,000km                 NaN    benzin  volkswagen   
freq         32424                 NaN     30107       10687   
mean           NaN            5.723360       NaN         NaN   
std            NaN            3.711984       NaN         NaN   
min            NaN            0.000000       NaN         NaN   
25%            NaN            3.000000       NaN         NaN   
50%            NaN            6.000000       NaN         NaN   
75%            NaN            9.000000       NaN         NaN   
max            NaN           12.000000       NaN         NaN   

       unrepaired_damage           ad_created  num_photos   postal_code  \
count              40171                50000     50000.0  50000.000000   
unique                 2                   76         NaN           NaN   
top                 nein  2016-04-03 00:00:00         NaN           NaN   
freq               35232                 1946         NaN           NaN   
mean                 NaN                  NaN         0.0  50813.627300   
std                  NaN                  NaN         0.0  25779.747957   
min                  NaN                  NaN         0.0   1067.000000   
25%                  NaN                  NaN         0.0  30451.000000   
50%                  NaN                  NaN         0.0  49577.000000   
75%                  NaN                  NaN         0.0  71540.000000   
max                  NaN                  NaN         0.0  99998.000000   

                  last_seen  
count                 50000  
unique                39481  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

We can see seller column has the same values as offer_tyoe column and num_photos has '0' value. We need to do further investigation of num_photos column to decide if we need to drop this column.

Any columns that have mostly one value that are candidates to be dropped Any examples of numeric data stored as text that needs to be cleaned. For example, Price and Oldmeter need to rmove non-numeric dtype and change datatype

In [12]:
#Remove some columns which have no values or same values
autos['num_photos'].value_counts()

0    50000
Name: num_photos, dtype: int64

As it shows, num_photos only has '0' value. Hence, I decide to drop this column as well.

In [13]:
autos.drop(columns=['seller','offer_type','num_photos'])

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_T�V_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine

In [14]:
#Change datatype to int for some colunmns so that I can do analysis later and remove outliers
autos['price']=autos['price'].str.replace('$','').str.replace(',','').astype(int)
autos['price'].head(5)

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int64

In [15]:
autos['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [16]:
#Find out the max and the min so I can decide if I need to remove the outlier
autos['price'].value_counts().sort_index(ascending=True)

0           1421
1            156
2              3
3              1
5              2
8              1
9              1
10             7
11             2
12             3
13             2
14             1
15             2
17             3
18             1
20             4
25             5
29             1
30             7
35             1
40             6
45             4
47             1
49             4
50            49
55             2
59             1
60             9
65             5
66             1
            ... 
151990         1
155000         1
163500         1
163991         1
169000         1
169999         1
175000         1
180000         1
190000         1
194000         1
197000         1
198000         1
220000         1
250000         1
259000         1
265000         1
295000         1
299000         1
345000         1
350000         1
999990         1
999999         2
1234566        1
1300000        1
3890000        1
10000000       1
11111111       2
12345678      

Min and Max price are $0 and $99,999,999, which is unreasonable number for used cars. It will impact on the mean and median in the following analysis. Hence, I need to remove outliers. Ebay is an auction website, so $1 makes sense here. However, look at the data carefully,we can see $999,999 is a big jump from  $350,000.So we will exclude any cars with price over $350,000 in the dataset.

In [17]:
autos=autos[autos['price'].between(1,350000)]
autos['price'].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

After removing the outliners, we can see the mean price is much more reasonable.
The next step is to explore odometer_km column. we will do the same steps.

In [19]:
autos['odometer_km'].unique().shape

(13,)

In [30]:
autos['odometer_km']=autos['odometer_km'].str.replace(',','').str.replace('km','').astype(int)
autos['odometer_km'].describe()

count     48565.000000
mean     125770.101925
std       39788.636804
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [31]:
autos['odometer_km'].value_counts().sort_index()

5000        836
10000       253
20000       762
30000       780
40000       815
50000      1012
60000      1155
70000      1217
80000      1415
90000      1734
100000     2115
125000     5057
150000    31414
Name: odometer_km, dtype: int64

No need to remove outliners,numbers are reasonable. 

In [40]:
#Check  registration year
autos['year']=autos['registration_year'][:4]
#Extract year only and add a new column 'Year' to the dataset

In [42]:
autos['registration_year'].value_counts().sort_index(ascending=False)


9999       3
9000       1
8888       1
6200       1
5911       1
5000       4
4800       1
4500       1
4100       1
2800       1
2019       2
2018     470
2017    1392
2016    1220
2015     392
2014     663
2013     803
2012    1310
2011    1623
2010    1589
2009    2085
2008    2215
2007    2277
2006    2670
2005    2936
2004    2703
2003    2699
2002    2486
2001    2636
2000    3156
        ... 
1964      12
1963       8
1962       4
1961       6
1960      23
1959       6
1958       4
1957       2
1956       4
1955       2
1954       2
1953       1
1952       1
1951       2
1950       3
1948       1
1943       1
1941       2
1939       1
1938       1
1937       4
1934       2
1931       1
1929       1
1927       1
1910       5
1800       2
1111       1
1001       1
1000       1
Name: registration_year, Length: 95, dtype: int64

In [44]:
autos=autos[autos['registration_year'].between(1986,2016)]
autos['registration_year'].value_counts(normalize=True)

2000    0.068853
2005    0.064053
1999    0.063202
2004    0.058970
2003    0.058883
2006    0.058250
2001    0.057508
2002    0.054236
1998    0.051552
2007    0.049676
2008    0.048323
2009    0.045487
1997    0.042564
2011    0.035408
2010    0.034666
1996    0.029954
2012    0.028580
1995    0.026769
2016    0.026616
2013    0.017519
2014    0.014464
1994    0.013723
1993    0.009272
2015    0.008552
1992    0.008072
1990    0.007570
1991    0.007396
1989    0.003796
1988    0.002945
1987    0.001571
1986    0.001571
Name: registration_year, dtype: float64

Registration date shows how old the car is. Rmove years before year 1986, we only focus on usage year within 30 years. The dataset is collected in year 2016, so any years later than 2016 is incorrect.

In [32]:
#Format date columns
autos.loc[0:5,['date_crawled','ad_created','last_seen']]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50
5  2016-03-21 13:47:45  2016-03-21 00:00:00  2016-04-06 09:45:21

In [37]:
#Extract year-month-date only and save to autos
autos['date_crawled']=autos['date_crawled'] .str[:10]
autos['ad_created']=autos['ad_created'].str[:10]
autos['last_seen']=autos['last_seen'].str[:10]


Data cleaning is finsihed at this step. Now I am going to analyse the dataset.I need to obtain insight informations from the data. 

I will answers below questions.
1, What date or period has the msot ad created?  
2, What brands are popular and how different average price are they.
3, Are there's any visible link with mean price and the average milage?
4, What is the most common brand/model combinations?
5, Does average price follow any patterns basd on the milage?
6, How much cheaper are cars with damage than their non-damaged counterparts?

In [34]:
#1,Explore the date columns 
autos['date_crawled'].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

In [35]:
autos['ad_created'].value_counts(normalize=True, dropna=False).sort_index()

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000041
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000062
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000041
2016-02-05    0.000041
2016-02-07    0.000021
2016-02-08    0.000021
2016-02-09    0.000021
2016-02-11    0.000021
2016-02-12    0.000041
2016-02-14    0.000041
2016-02-16    0.000021
2016-02-17    0.000021
2016-02-18    0.000041
2016-02-19    0.000062
2016-02-20    0.000041
2016-02-21    0.000062
                ...   
2016-03-09    0.033151
2016-03-10    0.031895
2016-03-11    0.032904
2016-03-12    0.036755
2016-03-13    0.017008
2016-03-14    0.035190
2016-03-15    0.034016
2016-03-16    0.030125
2016-03-17    0.031278
2016-03-18    0.013590
2016-03-19    0.033687
2016-03-20    0.037949
2016-03-21 

In [36]:
autos['last_seen'].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

After checking the date columns, Ebay site was crawled daily between March and April 2016. After normalizing data, it seems the distribution is normal.The date_crawled data provides us about any listing that it last saw. This helps us to determine on what day a listing was removed, with an assumption that the car was sold. Ad created date is between June,2015 and April,2016, most data lie in beteween March 2016 and April 2016.

In [45]:
#2,The next step is to explore variations across different car brands. We can find out what brands are popular and how different average price are they.
top_brands = autos['brand'].value_counts(normalize=True)
top_brands=top_brands[top_brands>0.05].index
#take brand names which are higher than 5% frequency rate
print(top_brands)

average_price_brand={}
for brand in top_brands:
    selected_rows=autos[autos['brand']==brand]
    mean_of_selected_brand=selected_rows['price'].mean()
    average_price_brand[brand]=int(mean_of_selected_brand)
sort_average_price_brand=sorted(average_price_brand.items(),key=lambda x:x[1],reverse=True)
print(sort_average_price_brand)

Index([u'volkswagen', u'bmw', u'opel', u'mercedes_benz', u'audi', u'ford'], dtype='object')
[('audi', 9352), ('mercedes_benz', 8495), ('bmw', 8332), ('volkswagen', 5364), ('ford', 3439), ('opel', 2953)]


In order to understand what band hold the best value in the used car listing,we analyzed average price for each brand by taking brands having over 5% of total listings.

We can conclude that Audi,Mercedes_benz and BMW are more expensive. Audi takes the first place,which has $857 more than the second place,Mercedesbenz.However,there are only $163 difference between the second place and the third place. Ford and Opel are less expensive and Volkswagen is in between.

In [46]:
#3,For the top 6 brands, let's use aggregation to understand the average mileage for those cars and if there's any visible link with mean price.
average_mileage={}
for brand in top_brands:
    selected_rows=autos[autos['brand']==brand]
    mean_of_mileage=selected_rows['odometer_km'].mean()
    average_mileage[brand]=int(mean_of_mileage)
sort_average_mileage=sorted(average_mileage.items(),key=lambda x:x[1],reverse=True)
print(sort_average_mileage)

[('bmw', 132738), ('mercedes_benz', 131110), ('opel', 129836), ('audi', 129143), ('volkswagen', 129116), ('ford', 125272)]


In [47]:
avg_price=pd.Series(average_price_brand)
#change dictionary to series
avg_price=pd.DataFrame(avg_price,columns=['Mean Price'])
#change series to dataframe
print(avg_price)

               Mean Price
audi                 9352
bmw                  8332
ford                 3439
mercedes_benz        8495
opel                 2953
volkswagen           5364


In [48]:
avg_mil=pd.Series(average_mileage)
avg_mil=pd.DataFrame(avg_mil,columns=['Mean Mileage'])
avg_price['Mean Mileage']=avg_mil['Mean Mileage']
avg_price=avg_price.sort_values(by='Mean Price',ascending=False)
print(avg_price)

               Mean Price  Mean Mileage
audi                 9352        129143
mercedes_benz        8495        131110
bmw                  8332        132738
volkswagen           5364        129116
ford                 3439        125272
opel                 2953        129836


Mileage does not impact price on all 6 brands. Compared the hightest mean price brand,Audi,and the lowest mean price brand, Opel, mean price has over $6000 difference while mean mileage of opeal is higher than Audi. Even thought Mercedes and BMW have a higher mean mileage than Volkswagen,Ford and Opel, their mean prices are still higher than those three.

In [49]:
#4,Explore what is the most common brand/model combinations?
autos['Brand/Model']=autos['brand']+'/'+autos['model']
autos['Brand/Model'].value_counts().head(5)

volkswagen/golf    3676
bmw/3er            2601
volkswagen/polo    1605
opel/corsa         1591
opel/astra         1348
Name: Brand/Model, dtype: int64

The most common brand/model combinations is volkswagen/golf, with 3676 unit. 

In [50]:
#5,Does average price follow any patterns basd on the milage?

#Devide milage into 3 groups.
group1=autos[autos['odometer_km'].between(0,50000)]
group2=autos[autos['odometer_km'].between(50000,100000)]
group3=autos[autos['odometer_km'].between(100000,150000)]

In [70]:
def group_avg_price(groupnum):
    avg_price=groupnum['price'].mean()
    return int(avg_price)
print('Group1 has average price',group_avg_price(group1))
print('Group2 has average price',group_avg_price(group2))
print('Group3 has average price',group_avg_price(group3))
    

('Group1 has average price', 15020)
('Group2 has average price', 9998)
('Group3 has average price', 4265)


As the milage increases, the average price decreases.
Milage lower than 50,000km,the average price is $15,020, milage between 50,000km and 100,000km,the average price is $9,998, milage between 100,000km and 150,000km, the average price is $4,265.

In [71]:
#5,How much cheaper are cars with damage than their non-damaged counterparts?
autos['unrepaired_damage'].value_counts(dropna=False)


nein    33324
NaN      8046
ja       4467
Name: unrepaired_damage, dtype: int64

In [72]:
#Change German to English
G_to_E={'nein':'No','ja':'Yes'}
autos['unrepaired_damage']=autos['unrepaired_damage'].map(G_to_E)
autos['unrepaired_damage'].unique()

array(['No', nan, 'Yes'], dtype=object)

In [74]:
withdamage=autos['unrepaired_damage']=='Yes'
damagecars=autos.loc[withdamage,:]
dc=damagecars.groupby('Brand/Model')['price'].mean().reset_index()
dc.rename(columns={'price':'dc Mean Price'},inplace=True)
print(dc)

                Brand/Model  dc Mean Price
0            alfa_romeo/147    1286.846154
1            alfa_romeo/156    1190.388889
2            alfa_romeo/159    7800.000000
3         alfa_romeo/andere    3548.625000
4                  audi/100    1166.500000
5                   audi/80     641.421053
6                   audi/a2    2333.333333
7                   audi/a3    2769.447761
8                   audi/a4    2943.366337
9                   audi/a5   21649.800000
10                  audi/a6    3201.715789
11                  audi/a8    3780.000000
12              audi/andere    5066.583333
13                  audi/q7   11950.000000
14                  audi/tt    7521.111111
15                  bmw/1er    6220.645161
16                  bmw/3er    2329.507042
17                  bmw/5er    3926.722772
18                  bmw/7er    5070.846154
19               bmw/andere   15149.500000
20                   bmw/i3     250.000000
21              bmw/m_reihe    8900.000000
22         

In [75]:
nodamage=autos['unrepaired_damage']=='No'
nodamagecars=autos.loc[nodamage,:]
ndc=nodamagecars.groupby('Brand/Model')['price'].mean().reset_index()
ndc.rename(columns={'price':'ndc Mean Price'},inplace=True)
print(ndc)

                Brand/Model  ndc Mean Price
0            alfa_romeo/145     2750.000000
1            alfa_romeo/147     2722.854545
2            alfa_romeo/156     1689.516667
3            alfa_romeo/159     6659.653846
4         alfa_romeo/andere     7401.850000
5         alfa_romeo/spider     6663.277778
6                  audi/100     1978.750000
7                  audi/200     2250.000000
8                   audi/80     1751.427083
9                   audi/90     1699.600000
10                  audi/a1    14793.827160
11                  audi/a2     3808.848485
12                  audi/a3     9263.882353
13                  audi/a4     8175.672451
14                  audi/a5    22454.473214
15                  audi/a6     9754.379139
16                  audi/a8    11933.750000
17              audi/andere    22828.326425
18                  audi/q3    28346.962963
19                  audi/q5    26827.660714
20                  audi/q7    22040.184211
21                  audi/tt    1

In [77]:
df=pd.merge(dc,ndc,how='left',left_on='Brand/Model',right_on='Brand/Model')
print(df)

                Brand/Model  dc Mean Price  ndc Mean Price
0            alfa_romeo/147    1286.846154     2722.854545
1            alfa_romeo/156    1190.388889     1689.516667
2            alfa_romeo/159    7800.000000     6659.653846
3         alfa_romeo/andere    3548.625000     7401.850000
4                  audi/100    1166.500000     1978.750000
5                   audi/80     641.421053     1751.427083
6                   audi/a2    2333.333333     3808.848485
7                   audi/a3    2769.447761     9263.882353
8                   audi/a4    2943.366337     8175.672451
9                   audi/a5   21649.800000    22454.473214
10                  audi/a6    3201.715789     9754.379139
11                  audi/a8    3780.000000    11933.750000
12              audi/andere    5066.583333    22828.326425
13                  audi/q7   11950.000000    22040.184211
14                  audi/tt    7521.111111    13743.041667
15                  bmw/1er    6220.645161    12016.3600

In [78]:
df.describe()

dc Mean Price  ndc Mean Price
count     237.000000      236.000000
mean     3325.168269     7076.947843
std      3831.988785     7231.732329
min       150.000000      593.625000
25%      1058.411765     2923.658654
50%      2079.200000     4918.037037
75%      3926.722772     8465.611191
max     24333.333333    74588.900000

In order to understand if damage significantly affects average price, I compare the average price difference between damage cars and non-damage cars by brand/model. Firstly, I have to seperate damaged cars and non-damage cars, then calculate the avaerage price for each brand/model in both damaged car group and non-damagebcar group. Lastly, we can compare the average price. 

Take bmw/3er and volkswagen/golf as examples, non-damage bmw/3er has an average price $6935.75,damaged bmw/3er has an average price $2329.51. Non-damage volkswagen/golf has an average price $6091.40,damaged volkswagen/golf has an average price $1858.73. As all data show above, we can conclude that there is a significant price difference between damgaged cars and nondamage cars.